### PR TITLE
Song Duration

### PR DESCRIPTION
**Pull request by cheezy64**
## Proposed Changes
Implements #45 
Screen Shot 2019-11-19 at 12 38 05 AM

Two entries added, one for backwards compatibility (no duration) and one with the new feature.

## Platforms
This pull request modifies: *(select all that apply)*
- [x] Client
- [x] Server

## Types of Changes
This pull request is contains: *(select all that apply)*
- [ ] Bug fixes *(non-breaking change which fixes an issue)*
- [x] New features *(non-breaking change which adds functionality)*
- [ ] Breaking changes *(fix or feature that would cause existing functionality to not work as expected)*

## Checklist
- [x] I have read the [contribution guidelines](https://github.com/lolPants/beatsaver-reloaded/blob/master/.github/CONTRIBUTING.md)
- [x] I have checked the changes adhere to the project's style guide and all tests pass
- [x] I have (to the best of my ability) checked for vulnerabilities, or any ways that my code could be abused and concluded that it is safe to run in production

## Further Comments
For the client, I wasn't sure of the best way to give more space for the "duration" icon.  So, I just made the image size a little bigger.  It's hacky so I'm open to suggestions.

For the server, an approximate duration can be calculated with the BPM and obstacle info.  It will be wildly off if there's 10+ seconds of no obstacles near the end.  So, I opted to use `ffprobe`.  My original intent was to pipe the buffer into `ffprobe -i pipe:0`, but it fails due to a limitation with the `ogg` file format.  So, an intermediate file was necessary.

### CODE REVIEW

File: server/src/routes/upload/parseBeatmap.ts [hunk 1]
```diff
@@ -261,3 +271,40 @@ const inspectFile = async (file: JSZip.JSZipObject) => {
...
+    'a:0',
+    '-show_entries',
+    'stream=duration',
+    '-of',
+    'default=noprint_wrappers=1:nokey=1',
+  ]
+
+  // Intermediate file is necessary since an ogg file can't be piped to ffprobe via stdin
+  //   https://stackoverflow.com/questions/3713148/cant-stream-ogg-from-ffmpeg-through-stdout
+  const intermediateFilename = `./${sourceFilename}`
```
Comment (lolPants) [inline server/src/routes/upload/parseBeatmap.ts]: This should be written to the system's temporary directory, ideally with the `hash` as a subdirectory to avoid race conditions with the same `song.egg`

File: server/src/routes/upload/parseBeatmap.ts [hunk 2]
```diff
@@ -1,4 +1,8 @@
 import { createHash } from 'crypto'
+import { promises as fs } from 'fs'
```
Comment (lolPants) [inline server/src/routes/upload/parseBeatmap.ts]: This is still experimental, use `~utils/fs`

File: server/src/mongo/models/Beatmap.ts [hunk 3]
```diff
@@ -103,6 +104,7 @@ const schema: Schema = new Schema({
       required: true,
       type: String,
     },
+    songDuration: { type: Number, default: 0 },
```
Comment (lolPants) [inline server/src/mongo/models/Beatmap.ts]: Some form of validation on this (ie cannot be negative) would be nice

File: client/src/ts/components/Beatmap/Statistics.tsx [hunk 4]
```diff
@@ -85,3 +101,18 @@ export const BeatmapStats: FunctionComponent<IProps> = ({
 const isFullMap: (map: IMapStats | IBeatmap) => map is IBeatmap = map => {
   return (map as IBeatmap).downloadURL !== undefined
 }
+
+const convertSecondsToTime: (duration: number) => string = duration => {
+  const hours = Math.trunc(duration / 3600)
+  const minutes = Math.trunc((duration % 3600) / 60)
+  const seconds = Math.trunc(duration % 60)
+
+  let formattedDuration = ''
```
Comment (lolPants) [inline client/src/ts/components/Beatmap/Statistics.tsx]: Avoid using `let`.
Try to rewrite with a more functional approach

File: server/src/routes/upload/parseBeatmap.ts [hunk 5]
```diff
@@ -87,6 +91,11 @@ export const parseBeatmap: (
     throw ERR_BEATMAP_AUDIO_INVALID
   }
 
+  const songDuration = await getDurationInSeconds(
+    `${infoJSON._songAuthorName}-${infoJSON._songName}`,
```
Comment (lolPants) [inline server/src/routes/upload/parseBeatmap.ts]: See my other comment about temporary directories. Consider passing in `infoJSON._songFilename`, and the map's `hash` (you'll need to move the function call after the hash calculation) instead.

File: client/src/sass/_beatmap-result.scss [hunk 6]
```diff
@@ -1,5 +1,5 @@
 $margin: 14px;
-$image-size: 180px;
+$image-size: 190px;
```
Comment (lolPants) [inline client/src/sass/_beatmap-result.scss]: Why was this changed?

Comment (cheezy64) [reply client/src/sass/_beatmap-result.scss]: It was a hack to give more space around the last `Statistic` item.  I'll revert the changes to this file, but can you give me some guidance on how you'd like to handle this?

File: server/src/routes/upload/parseBeatmap.ts (hunk 2 above)
Comment (cheezy64) [reply server/src/routes/upload/parseBeatmap.ts]: Whoops, I didn't see this module.  Fixed

File: server/src/mongo/models/Beatmap.ts (hunk 3 above)
Comment (cheezy64) [reply server/src/mongo/models/Beatmap.ts]: Good recommendation.  Some validation is always nice

File: server/src/routes/upload/parseBeatmap.ts (hunk 1 above)
Comment (cheezy64) [reply server/src/routes/upload/parseBeatmap.ts]: I decided to use `tmp-promise` package for this.  It doesn't add any extra dependencies that the server doesn't already have, and it provides uniqueness and file cleanup.

File: client/src/ts/components/Beatmap/Statistics.tsx (hunk 4 above)
Comment (cheezy64) [reply client/src/ts/components/Beatmap/Statistics.tsx]: I took a stab at it.  If you had something else in mind, please provide an example

File: server/src/routes/upload/parseBeatmap.ts (hunk 5 above)
Comment (cheezy64) [reply server/src/routes/upload/parseBeatmap.ts]: Please see above comment.  I opted to focus on uniqueness, but still passed in the `hash` (even though it's not necessary), but it might be useful for identifying the file for future debugging

File: client/src/ts/components/Beatmap/Statistics.tsx (hunk 4 above)
Comment (lolPants) [reply client/src/ts/components/Beatmap/Statistics.tsx]: Refactored code looks good 👍
